sentry-android-core/src/main/java/io/sentry/android/core/AndroidProfiler.java

```diff
@@ -0,0 +1,360 @@
+        final long endNanos,
+        final long endCpuMillis,
+        final boolean didTimeout,
+        final @NotNull File traceFile,
+        final @NotNull Map<String, ProfileMeasurement> measurementsMap) {
+      this.endNanos = endNanos;
+      this.traceFile = traceFile;
+      this.endCpuMillis = endCpuMillis;
+      this.measurementsMap = measurementsMap;
+      this.didTimeout = didTimeout;
+    }
+  }
+
+  /**
+   * This appears to correspond to the buffer size of the data part of the file, excluding the key
+   * part. Once the buffer is full, new records are ignored, but the resulting trace file will be
+   * valid.
+   *
+   * <p>30 second traces can require a buffer of a few MB. 8MB is the default buffer size for
+   * [Debug.startMethodTracingSampling], but 3 should be enough for most cases. We can adjust this
+   * in the future if we notice that traces are being truncated in some applications.
+   */
+  private static final int BUFFER_SIZE_BYTES = 3_000_000;
+
+  private static final int PROFILING_TIMEOUT_MILLIS = 30_000;
+  private long transactionStartNanos = 0;
+  private final @NotNull File traceFilesDir;
+  private final int intervalUs;
+  private @Nullable Future<?> scheduledFinish = null;
+  private @Nullable File traceFile = null;
+  private @Nullable String frameMetricsCollectorId;
+  private volatile @Nullable ProfileEndData timedOutProfilingData = null;
+  private final @NotNull SentryFrameMetricsCollector frameMetricsCollector;
+  private final @NotNull ArrayDeque<ProfileMeasurementValue> screenFrameRateMeasurements =
+      new ArrayDeque<>();
+  private final @NotNull ArrayDeque<ProfileMeasurementValue> slowFrameRenderMeasurements =
+      new ArrayDeque<>();
+  private final @NotNull ArrayDeque<ProfileMeasurementValue> frozenFrameRenderMeasurements =
+      new ArrayDeque<>();
+  private final @NotNull Map<String, ProfileMeasurement> measurementsMap = new HashMap<>();
+  private final @NotNull SentryAndroidOptions options;
+  private final @NotNull BuildInfoProvider buildInfoProvider;
+  private boolean isRunning = false;
+
+  public AndroidProfiler(
+      final @NotNull String tracesFilesDirPath,
+      final int intervalUs,
+      final @NotNull SentryFrameMetricsCollector frameMetricsCollector,
+      final @NotNull SentryAndroidOptions sentryAndroidOptions,
+      final @NotNull BuildInfoProvider buildInfoProvider) {
+    this.traceFilesDir =
+        new File(Objects.requireNonNull(tracesFilesDirPath, "TracesFilesDirPath is required"));
+    this.intervalUs = intervalUs;
+    this.options = Objects.requireNonNull(sentryAndroidOptions, "SentryAndroidOptions is required");
+    this.frameMetricsCollector =
+        Objects.requireNonNull(frameMetricsCollector, "SentryFrameMetricsCollector is required");
+    this.buildInfoProvider =
+        Objects.requireNonNull(buildInfoProvider, "The BuildInfoProvider is required.");
+  }
+
+  @SuppressLint("NewApi")
+  public @Nullable ProfileStartData start() {
```

Comment:
I guess we can't always be certain on which thread profiling is started / stopped?
How about marking start(), endCollection() and close() `synchronized`? This way e.g.
 *  `isRunning` would at least be guarded and parallel calls to e.g. `start()` wouldn't cause any invalid state.
 * `endCollection()` wouldn't be executed in parallel in case it's called by scheduleFinish + manually at the same time

Reply:
Make sense to me, I've updated the methods.